src/cpp/src/scheduler.hpp

```diff
@@ -354,11 +366,15 @@ class Scheduler {
                     Sequence::Ptr sequence = (*sequence_group)[0];
                     uint64_t seq_id = sequence->get_id();
 
-                    // allocate KV blocks
-                    m_block_manager.allocate(seq_id, num_required_blocks);
                     // and schedule tokens
                     sequence_group->schedule_tokens(sequence_len);
 
+                    // allocate KV blocks
+                    if (sequence_group->get_num_processed_tokens() == 0)
+                        m_block_manager.allocate(sequence, num_required_blocks, sequence_group->get_prompt_ids());
+                    else 
+                        m_block_manager.append_slots(sequence_group);
```

Comment:
why do we need such condition based on get_num_processed_tokens ?

Reply:
`append_slots()` is needed, because after blocks restoring we might not need to allocate a new block, this situation happens only if `sequence_group->get_num_processed_tokens() > 0`. 

But `append_slots()` can actually be used in both cases, so I removed the condition and used `append_slots()`.